src/http2_frame_headers.erl

```diff
@@ -91,3 +83,74 @@ from_frames([], Acc) ->
     Acc;
 from_frames([{#frame_header{type=?CONTINUATION},#continuation{block_fragment=BF}}|Continuations], Acc) ->
     from_frames(Continuations, <<Acc/binary,BF/binary>>).
+
+-spec split(Binary::binary(),
+            MaxFrameSize::pos_integer()) ->
+                   [binary()].
+split(Binary, MaxFrameSize) ->
+    split(Binary, MaxFrameSize, []).
+
+-spec split(Binary::binary(),
+            MaxFrameSize::pos_integer(),
+            [binary()]) ->
+                   [binary()].
+split(Binary, MaxFrameSize, Acc)
+  when byte_size(Binary) =< MaxFrameSize ->
+    lists:reverse([Binary|Acc]);
+split(Binary, MaxFrameSize, Acc) ->
+    <<NextFrame:MaxFrameSize/binary,Remaining/binary>> = Binary,
+    split(Remaining, MaxFrameSize, [NextFrame|Acc]).
+
+%% Now build frames.
+%% The first will be a HEADERS frame, followed by CONTINUATION
+%% If EndStream, that flag needs to be set on the first frame
+%% ?FLAG_END_HEADERS needs to be set on the last.
+%% If there's only one, it needs to be set on both.
+-spec build_frames(StreamId :: stream_id(),
+                   Chunks::[binary()],
+                   EndStream::boolean()) ->
+                          [frame()].
+build_frames(StreamId, [FirstChunk|Rest], EndStream) ->
+    Flag = case EndStream of
+               true ->
+                   ?FLAG_END_STREAM;
+               false ->
+                   0
+           end,
+    HeadersFrame =
+        { #frame_header{
+             type=?HEADERS,
+             flags=Flag,
+             length=byte_size(FirstChunk),
+             stream_id=StreamId},
+          #headers{
+             block_fragment=FirstChunk}},
+    [{LastFrameHeader, LastFrameBody}|Frames] =
+        build_frames_(StreamId, Rest, [HeadersFrame]),
+    NewLastFrame = {
+      LastFrameHeader#frame_header{
+        flags=LastFrameHeader#frame_header.flags bor ?FLAG_END_HEADERS
+       },
+      LastFrameBody},
+
+    lists:reverse([NewLastFrame|Frames]).
+
+-spec build_frames_(StreamId::stream_id(),
+                    Chunks::[binary()],
+                    Acc::[frame()])->
+                           [frame()].
+build_frames_(_StreamId, [], Acc) ->
```

Comment:
I'd still have it broken up into different functions because the initial caller shouldn't be responsible for setting up the accumulator